### PR TITLE
[k8s] Map possible Pod/Container status to ModuleStatus

### PR DESCRIPTION
**Pull request by ggjjj**
1. Changing the Pod status to module status conversion for Pending/Failed/Unknown status. 
2. Added K8s Environment and K8s Provider to report the module status to IOT hub instead of the docker environment.
3. Updated the unit tests to handle all the scenarios of Pod Phase and container status.
4. Report module runtime details to IOT hub instead of docker runtime info for module status and module description.
5. Better report of module status when there is no container schedule as pod failed with volume mounts or PVC claims

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                }
+
+                if (!moduleSet.Modules.TryGetValue(dockerRuntimeInfo.Name, out IModule configModule) || !(configModule is DockerModule dockerModule))
+                {
+                    dockerModule = new DockerModule(dockerRuntimeInfo.Name, string.Empty, ModuleStatus.Unknown, Core.RestartPolicy.Unknown, new DockerConfig(Constants.UnknownImage, new CreateContainerParameters()), ImagePullPolicy.OnCreate, new ConfigurationInfo(), null);
+                }
+
+                Option<ModuleState> moduleStateOption = await this.moduleStateStore.Get(moduleRuntimeInfo.Name);
+                ModuleState moduleState = moduleStateOption.GetOrElse(new ModuleState(0, moduleRuntimeInfo.ExitTime.GetOrElse(DateTime.MinValue)));
+
```
Comment (darobs) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs]: In Kubernetes mode, I'm 99.44% certain the module store is not written to - this is also Health Restart Planner stuff.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                            dockerModule.ImagePullPolicy,
+                            dockerModule.ConfigurationInfo,
+                            dockerModule.Env);
+                        break;
+                }
+
+                modules.Add(module);
+            }
+
+            return new ModuleSet(modules.ToDictionary(m => m.Name, m => m));
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs]:
```suggestion
            return ModuleSet.Create(modules.ToArray());
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/KubernetesModule.cs [hunk 3]
```diff
@@ -318,8 +318,8 @@ protected override void Load(ContainerBuilder builder)
                         var moduleStateStore = await c.Resolve<Task<IEntityStore<string, ModuleState>>>();
                         var restartPolicyManager = c.Resolve<IRestartPolicyManager>();
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Service/modules/KubernetesModule.cs]:
```suggestion
```
You don't need this variable

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs (hunk 1 above)
Comment (ggjjj) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs]: Today we still the module store to report the Last Exit time and restart count. For now, we don't report these info in K8s.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/PodExtensions.cs [hunk 4]
```diff
@@ -43,25 +43,67 @@ static Option<V1ContainerStatus> GetContainerByName(string name, V1Pod pod)
...
+                                return containerStatus.Map(c =>
+                                {
+                                    if (c.State.Waiting != null)
+                                    {
+                                        return new ReportedModuleStatus(ModuleStatus.Backoff, $"Module in Back-off because of the reason: {c.State.Waiting.Reason}");
+                                    }
+                                    else if (c.State.Terminated != null)
+                                    {
+                                        if (c.State.Terminated.ExitCode != 0)
+                                            return new ReportedModuleStatus(ModuleStatus.Failed, $"Module Failed becasue of the reason: {c.State.Terminated.Reason}");
```
Comment (darobs) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/PodExtensions.cs]: >becasue [](start = 113, length = 7)

typo

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                    Events.InvalidModuleType(moduleRuntimeInfo);
+                    continue;
+                }
+
+                if (!moduleSet.Modules.TryGetValue(dockerRuntimeInfo.Name, out IModule configModule) || !(configModule is DockerModule dockerModule))
+                {
+                    dockerModule = new DockerModule(dockerRuntimeInfo.Name, string.Empty, ModuleStatus.Unknown, Core.RestartPolicy.Unknown, new DockerConfig(Constants.UnknownImage, new CreateContainerParameters()), ImagePullPolicy.OnCreate, new ConfigurationInfo(), null);
+                }
+
+                Option<ModuleState> moduleStateOption = await this.moduleStateStore.Get(moduleRuntimeInfo.Name);
```
Comment (dmolokanov) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs]: Do we want to delete these 2 lines?

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs [hunk 6]
```diff
@@ -173,48 +173,47 @@ public async Task ReturnsLastKnowModuleState()
 
             Dictionary<string, V1Pod> modified = BuildPodList();
 
-            DateTime agentStartTime = new DateTime(2019, 6, 11);
+            string agentWaitingReason = "CrashBackLoopOff";
```
Comment (dmolokanov) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs]: Could you possibly split this monstrous test in several once? This test violates one of the most important principles: The [Single Responsibility](https://en.wikipedia.org/wiki/Single_responsibility_principle) principle. What is the main thing that verifies this test? We can discuss it offline.

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs [hunk 7]
```diff
@@ -3,21 +3,25 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test
 {
     using System;
     using System.Collections.Generic;
+    using System.ComponentModel;
     using System.IO;
     using System.Linq;
     using System.Text;
     using System.Threading;
     using System.Threading.Tasks;
+    using Akka.Streams.Implementation;
```
Comment (dmolokanov) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs]: looks like all new imports are redundand. resharper/rider indicates this ;)

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs [hunk 8]
```diff
@@ -160,209 +164,198 @@ public async Task ConvertsPodsToModules()
...
-            modified["edgehub"].Status.ContainerStatuses[0].State.Waiting = null;
-            modified["edgehub"].Status.ContainerStatuses[0].State.Terminated = new V1ContainerStateTerminated(0, finishedAt: DateTime.Parse("2019-06-12T16:13:07Z"), startedAt: DateTime.Parse("2019-06-12T16:11:22Z"), reason: edgehubTerminatedReason);
-
-            modified["simulatedtemperaturesensor"].Status.Phase = "Running";
-            modified["simulatedtemperaturesensor"].Status.ContainerStatuses[1].State.Running = new V1ContainerStateRunning(startedAt: DateTime.Parse("2019-06-12T16:11:22Z"));
-            modified["simulatedtemperaturesensor"].Status.ContainerStatuses[1].State.Waiting = null;
-            modified["simulatedtemperaturesensor"].Status.ContainerStatuses[1].State.Terminated = null;
-
-            foreach (V1Pod pod in modified.Values)
+            // module name "module-a"
```
Comment (dmolokanov) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs]: it's better to not leave this type of comments

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs [hunk 9]
```diff
@@ -160,209 +164,198 @@ public async Task ConvertsPodsToModules()
             }
         }
 
-        [Fact]
-        public async Task ReturnsLastKnowModuleState()
+        public static V1Pod CreatePodInPhaseWithContainerStatus(string podPhase, V1ContainerState containerState)
```
Comment (dmolokanov) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs]: Alternatively, this method can look like 
```csharp
public static V1Pod CreatePodInPhaseWithContainerStatus(string podPhase, V1ContainerState containerState)
    => new V1Pod 
    {
        ....
    };
```

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs [hunk 10]
```diff
@@ -160,209 +164,198 @@ public async Task ConvertsPodsToModules()
...
             {
-                if (string.Equals("edgeAgent", i.Name))
+                new object[]
                 {
-                    Assert.Equal(ModuleStatus.Backoff, i.ModuleStatus);
-                    Assert.Equal($"Module in Back-off because of the reason: {agentWaitingReason}", i.Description);
-                }
-                else if (string.Equals("edgeHub", i.Name))
+                    CreatePodInPhaseWithContainerStatus("Running", new V1ContainerState(waiting: new V1ContainerStateWaiting("Waiting", "CrashBackLoopOff"))),
+                    $"Module in Back-off reason: CrashBackLoopOff",
```
Comment (dmolokanov) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Kubernetes.Test/KubernetesRuntimeInfoProviderTest.cs]: `$` means that the following string is [interpolated](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) one and it's not needed here and below.

**Pull request merged**